packages/commonwealth/server/routes/upgradeMember.ts

```diff
@@ -14,120 +14,78 @@ export const Errors = {
   MustHaveAdmin: 'Communities must have at least one admin',
 };
 
-const ValidRoles = ['admin', 'moderator', 'member'];
+export const upgradeMemberValidation = [
+  body('new_role').custom(async (new_role) => {
+    if (!new_role || !isRole(new_role)) throw new AppError(Errors.InvalidRole);
+  }),
+  body('address').isString().withMessage(Errors.InvalidAddress),
+];
 
 const upgradeMember = async (
   models: DB,
   req: Request,
   res: Response,
   next: NextFunction
 ) => {
+  const errors = validationResult(req).array();
+  if (errors.length !== 0) {
+    return next(new AppError(errors[0].msg));
+  }
+
+  if (!req.user) return next(new AppError(Errors.NotLoggedIn));
+
   const chain = req.chain;
   const { address, new_role } = req.body;
-  if (!address) return next(new AppError(Errors.InvalidAddress));
-  if (!new_role) return next(new AppError(Errors.InvalidRole));
-  if (!req.user) return next(new AppError(Errors.NotLoggedIn));
+
+  // Get address ids of user making request to assert admin status
   const requesterAddresses = await req.user.getAddresses();
   const requesterAddressIds = requesterAddresses
     .filter((addr) => !!addr.verified)
     .map((addr) => addr.id);
-  const requesterAdminRoles = await findAllRoles(
-    models,
-    { where: { address_id: { [Op.in]: requesterAddressIds } } },
-    chain.id,
-    ['admin']
-  );
 
-  if (requesterAdminRoles.length < 1 && !req.user.isAdmin)
-    return next(new AppError(Errors.MustBeAdmin));
-  const memberAddress = await models.Address.findOne({
+  const addresses = await models.Address.findAll({
     where: {
-      address,
       chain: chain.id,
+      [Op.or]: {
+        id: { [Op.in]: requesterAddressIds },
+        address: address,
+      },
     },
   });
-  if (!memberAddress) return next(new AppError(Errors.NoMember));
-  const roles = await findAllRoles(
-    models,
-    { where: { address_id: memberAddress.id } },
-    chain.id
-  );
-  if (!roles) return next(new AppError(Errors.NoMember));
 
-  // There should only be one role per address per chain/community
-  const member = await findOneRole(
-    models,
-    { where: { address_id: memberAddress.id } },
-    chain.id
+  const targetAddress = addresses.find((a) => address === a.address);
+
+  // check if address provided exists
+  if (!targetAddress) return next(new AppError(Errors.NoMember));
+
+  const allCommunityAdmin = addresses.filter((m) => m.role === 'admin');
```

Comment:
A lot of this logic should be cleaned up / done at the DB level rather than in node, but again not 100% necessary if it's working.

Reply:
The issue is that we need to get multiple different categories for each Addresses query (list of admins, targetAddress, permissions of requesting addresses). So instead of making a separate query for each category, I just make one query and then filter them out via node. I think that it is more efficient to do it this way (this is the reason I rewrote this API instead of just replacing with shim).

Although I guess it is not clear without hard statistics. Maybe it is more efficient to make multiple queries for the specific categories instead of filtering them out via node.

Let me know what you think, then I can revert this file.

Reply:
This route isn't high priority to optimize, because it doesn't get used very often. Marking resolved.